File: src/common/Bank.cpp
#include "Bank.h"

namespace Management {
  Bank::Bank(int N) : price_level(2), gen(std::random_device().operator()()) {
    probOfPriceChange = {
            {1. / 3,  1. / 3,  1. / 6, 1. / 12, 1. / 12},
            {1. / 4,  1. / 3,  1. / 4, 1. / 12, 1. / 12},
            {1. / 12, 1. / 4,  1. / 3, 1. / 4,  1. / 12},
            {1. / 12, 1. / 12, 1. / 4, 1. / 3,  1. / 4},
            {1. / 12, 1. / 12, 1. / 6, 1. / 3,  1. / 3}
    };
    priceLevelInfo = {
            {1.0, 800, 3.0, 6500},
            {1.5, 650, 2.5, 6000},
            {2.0, 500, 2.0, 5500},
            {2.5, 400, 1.5, 5000},
            {3.0, 300, 1.0, 4500}
    };
    materials = floor(N * priceLevelInfo[price_level].UMat);
    product = floor(N * priceLevelInfo[price_level].UProd);
  }

  int Bank::getPriceLevel() const {
    return price_level + 1;
  }

  void Bank::setNewPriceLevel(int N) {
    price_level = probOfPriceChange[price_level](gen);
    materials = floor(N * priceLevelInfo[price_level].UMat);
    product = floor(N * priceLevelInfo[price_level].UProd);
  }

  DemandOffer Bank::getInfo() const {
    return priceLevelInfo[price_level];
  }

  std::vector<int> Bank::requestOfMat(const std::vector<Demand> &demand) {
    std::vector<int> accepted;
    std::vector<int> ord(demand.size());
    for (int i = 0; i < (int) demand.size(); i++)
      ord[i] = i;
    std::sort(ord.begin(), ord.end(), [&demand](int a, int b) { return demand[a] < demand[b]; });
    for (auto i : ord) {
      if (demand[i].Price < getInfo().MinPrice)
        break;
      if (demand[i].UMat > materials)
        continue;
      materials -= demand[i].UMat;
      accepted.push_back(i);
    }
    return accepted;
  }

  std::vector<int> Bank::requestOfProd(const std::vector<Offer> &offer) {
    std::vector<int> accepted;
    std::vector<int> ord(offer.size());
    for (int i = 0; i < (int) offer.size(); i++)
      ord[i] = i;
    std::sort(ord.begin(), ord.end(), [&offer](int a, int b) { return offer[a] < offer[b]; });
    for (auto i : ord) {
      if (offer[i].Price > getInfo().MaxPrice)
        break;
      if (offer[i].UProd > product)
        continue;
      product -= offer[i].UProd;
      accepted.push_back(i);
    }
    return accepted;
  }

  bool Demand::operator<(const Demand &x) const {
    return Price > x.Price || (Price == x.Price && Priority < x.Priority);
  }

  bool Offer::operator<(const Offer &x) const {
    return Price < x.Price || (Price == x.Price && Priority < x.Priority);
  }
}

File: Assets/Scripts/Model/Bank.cs
using UnityEngine;
using System.Collections.Generic;
using System.Collections;
using System;

namespace Management
{
    public struct Demand : IComparable<Demand>
    {
        public int Price { get; }
        public int UMat { get; }
        public int Priority { get; }

        public Demand(int price, int mat, int pr)
        {
            this.Price = price;
            this.UMat = mat;
            this.Priority = pr;
        }

        public int CompareTo(Demand other)
        {
            if (this.Price.CompareTo(other.Price) != 0)
                return -this.Price.CompareTo(other.Price);
            else
            {
                if (this.Priority.CompareTo(other.Priority) != 0)
                    return this.Priority.CompareTo(other.Priority);
                else
                    return this.UMat.CompareTo(other.UMat);

            }
        }
    }

    public struct Offer : IComparable<Offer>
    {
        public int Price { get; }
        public int UProd { get; }
        public int Priority { get; }

        public Offer(int price, int prod, int pr)
        {
            this.Price = price;
            this.UProd = prod;
            this.Priority = pr;
        }

        public int CompareTo(Offer other)
        {
            if (this.Price.CompareTo(other.Price) != 0)
                return this.Price.CompareTo(other.Price);
            else
            {
                if (this.Priority.CompareTo(other.Priority) != 0)
                    return this.Priority.CompareTo(other.Priority);
                else
                    return this.UProd.CompareTo(other.UProd);
            }
        }
    }

    public struct DemandOffer
    {
        public double UMat { get; } //units of material per one.
        public int MinPrice { get; }//minimal price of mat.

        public double UProd { get; }//units of product per one.
        public int MaxPrice { get; }//maximal price of prod.

        public DemandOffer(double mat, int minPr, double prod, int maxPr)
        {
            this.UMat = mat;
            this.MinPrice = minPr;
            this.UProd = prod;
            this.MaxPrice = maxPr;
        }
    }

    public class Bank
    {
        public static float[,] ProbOfPriceChange { get; } =
        {
            {1f/3f, 1f/3f, 1f/6f, 1f/12f, 1f/12f},
            {1f/4f, 1f/3f, 1f/4f, 1f/12f, 1f/12f},
            {1f/12f, 1f/4f, 1f/3f, 1f/4f, 1f/12f},
            {1f/12f, 1f/12f, 1f/4f, 1f/3f, 1f/4f},
            {1f/12f, 1f/12f, 1f/6f, 1f/3f, 1f/3f}
        };

        public static DemandOffer[] PriceLevelInfo { get; } =
        {
            new DemandOffer(1.0f, 800, 3.0f, 6500),
            new DemandOffer(1.5f, 650, 2.5f, 6000),
            new DemandOffer(2.0f, 500, 2.0f, 5500),
            new DemandOffer(2.5f, 400, 1.5f, 5000),
            new DemandOffer(3.0f, 300, 1.0f, 4500)
        };

        private int _priceLevel;
        public int PriceLevel => _priceLevel; 

        private int _materials;
        private int _product;

        private DemandOffer[] _report;
        public DemandOffer[] Report => _report;

        public Bank(int N)
        {
            _priceLevel = 2;
            _product = (int)Math.Floor(PriceLevelInfo[PriceLevel].UProd * N);
            _materials = (int)Math.Floor(PriceLevelInfo[PriceLevel].UMat * N);
        }

        public void SetNewPriceLevel(int N)
        {
            float Rem = 1.0f;
            for (int i = 0; i < ProbOfPriceChange.GetLength(_priceLevel); i++)
            {
                if (UnityEngine.Random.Range(0.0f, Rem) <= ProbOfPriceChange[_priceLevel, i])
                {
                    _priceLevel = i;
                    break;
                }
                Rem -= ProbOfPriceChange[_priceLevel, i];
            }
            _materials = (int)Math.Floor(PriceLevelInfo[_priceLevel].UMat * N);
            _product = (int)Math.Floor(PriceLevelInfo[_priceLevel].UProd * N);
        }

        public DemandOffer GetInfo => PriceLevelInfo[_priceLevel];

        public List<Demand> RequestOfMat(List<Demand> demands) //returns array of getting mat.
        {
            List<Demand> res = new List<Demand>();
            demands.Sort();
            foreach (Demand d in demands)
            {
                int got = Math.Min(_materials, d.UMat);
                if (d.Price < GetInfo.MinPrice)
                    got = 0;
                _materials -= got;
                res.Add(new Demand(d.Price, got, d.Priority));
            }
            return res;
        }

        public List<Offer> RequestOfProd(List<Offer> offers) //returns array of getting prod.
        {
            List<Offer> res = new List<Offer>();
            offers.Sort();
            foreach (Offer o in offers)
            {
                int got = Math.Min(_product, o.UProd);
                if (o.Price > GetInfo.MaxPrice)
                    got = 0;
                _product -= got;
                res.Add(new Offer(o.Price, got, o.Priority));
            }
            return res;
        }
    }
}